docs/contributor/adr/0010-geoblocking-draft.md

```diff
@@ -0,0 +1,387 @@
+- etag - used when checking for updates
+- lastUpdateCheckTime - the time when the list update has been performed (successfully) 
+- lastUpdateTime - the time when the list has been updated
+
+The above Config Map should only be used by the ip-auth service for caching/fallback purpose, it can't be treated as an official contract. It may be replaced with some other solution at any time without further notice.
+
+Example of an IP range allow/block list cache:
+
+```
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: ip-range-list-cache
+data:
+  version: ip-list-20240912-2200
+  etag: PeJwSe7ncNQchtGbkCPJ
+  lastUpdateCheckTime: "2024-09-17T15:28:50Z"
+  lastUpdateTime: "2024-09-17T14:28:50Z"
+  allow: |
+    - 192.0.2.10/32
+    - 2001:0002:6c::430
+  block: |
+    - 192.0.2.0/24
+    - 198.51.100.0/24
+    - 203.0.113.0/24
+    - 2001:0002::/48
+```
+
+![SAP internal service](../../assets/geoblocking-SAP-service.svg)
+
+### Geoblocking CR and controller
+
+In order to ensure reliability and configurability a new Geoblocking Custom Resource and a new Geoblocking Controller is introduced. The controller is responsible for:
+- managing the ip-auth service deployment
+- managing authorization policy that plugs ip-auth authorizer to all incomming requests
+- performing configuration checks (like external traffic policy)
+- reporting geoblocking state
+
+There should be only one Geoblocking resource. The controller should raise an issue if there are more of them similarly to [APIGateway CR](https://github.com/kyma-project/api-gateway/blob/05ed57f3299f42c2565ed1b28b84dee808a5213b/controllers/operator/apigateway_controller.go#L100).
+
+![Geoblocking controller](../../assets/geoblocking-cr-controller.svg)
+
+#### CR examples
+
+##### Usage of SAP internal service
+
+```yaml
+apiVersion: geoblocking.kyma-project.io/v1alpha1
+kind: GeoBlocking
+metadata:
+  name: geoblocking
+  namespace: kyma-system
+spec:
+  gbService:
+    secret: "gb-secret"
+    tokenUrl: https://oauth.example.com/oauth2/token
+    ipRange:
+      url: https://lists.example.com/lists/some-list
+      refreshInSeconds: 3600
+    events:
+      url: https://events.example.com/events
+      lobId: "kyma"
+      productId: "some-product"
+      systemId: "some-system"
+```
+
+##### IP range allow/block list populated by customer
+```yaml
+apiVersion: geoblocking.kyma-project.io/v1alpha1
+kind: GeoBlocking
+metadata:
+    name: geoblocking
+    namespace: kyma-system
+spec:
+  ipRange:
+    configMap: "my-own-ip-range-list"
+```
+
+### Technical details
+
+#### Policy download optimization
+
+In order to reduce unnecessary updates of the list of blocked IP ranges the ip-auth service should use ETag mechanism, which is supported by the SAP internal service.
+
+The local copy (stored in a Config Map) should be used to optimize the amount of connections to the SAP internal service:
+- if the local copy contains a newer version (lastUpdateTime) -> load it
```

Comment:
> another pod already loaded the latest version
Yes, this is what I meant, I'll adapt it to be more descriptive :)

>a pod would always check first if the config map has a newer version than the one in memory
Yes, exactly. It is all about utilizing more the local copy instead of flooding SAP service with requests from all atomic pods. So the idea was that any pod checks for the update once per some time and either updates the whole list (if there is an update) or updates the lastUpdateTime to indicate that the check was performed and the list was still up to date.
And yes, it is theoretically possible that there is a new policy list in the local copy and even newer policy list in the geoblocking service, but in normal operation it is rather unlikely.
So I would say that it is an implementation detail - the code that looks at local copy and the one that looks at the SAP service may be completely separate, triggered by different timers, etc.

Reply:
I adapted most of these points to be more precise, like:
```if the local copy contains a newer version of the policy list (lastUpdateTime attribute) -> it means that another pod has already downloaded it from the SAP internal service, so load the policy list from the local copy```